feat: introduce EntityFilter class with support for and/or filters

danieljbruce opened the pull request:
This code will enable the composite OR filters (much like the same way the composite AND filters work).

danieljbruce committed:
or filters interface

danieljbruce committed:
implementation of new filter

danieljbruce committed:
Remove duplicate code

danieljbruce committed:
Parse with type guard

danieljbruce committed:
Remove the newFilter variable

danieljbruce committed:
Add filter and enable chaining

danieljbruce committed:
test add filter

danieljbruce committed:
Add another unit test

danieljbruce committed:
Add system-test stubs

danieljbruce committed:
Add system-tests for the OR filter

danieljbruce committed:
Move things around

danieljbruce committed:
Merge branch 'main' into or-filters

danieljbruce committed:
Added a unit test

danieljbruce committed:
Merge branch 'or-filters' of https://github.com/danieljbruce/nodejs-datastore into or-filters

danieljbruce committed:
Add a unit test for OR filters

danieljbruce committed:
Just use filter method

danieljbruce commented on system-test/datastore.ts:
```diff
@@ -798,6 +799,83 @@ describe('Datastore', () => {
+            ])
+          );
+        const [entities] = await datastore.runQuery(q);
+        assert.strictEqual(entities!.length, 6);
```
This will change.

danieljbruce committed:
new warning test

danieljbruce committed:
Revert "new warning test"

This reverts commit 37400a6d3a9d7385a0203db70949f037e4c35de4.

danieljbruce committed:
Now removes deprecation warning properly

danieljbruce committed:
Add a test for new warning

danieljbruce committed:
Add setAncestor

Adds a new setAncestor method for ensuring only one ancestor is set for the query at a time. This will avoid errors that result because of setting multiple ancestors. Also deprecate hasAncestor because it will lead to warnings like this. Add parser logic to use the value provided in setAncestor for query sent to backend.

danieljbruce committed:
Basic unit tests for setAncestor

Added tests for the query proto, one to make sure the query structure is right when setting ancestor once and one to make sure the query structure is right when setting ancestor twice. Also added a unit test to make sure that ancestor is set the right way internally when using setAncestor.

danieljbruce committed:
change expected result for OR query

This code change adjusts the expected result for running an OR query. Old result used to correspond with AND, but now corresponds to OR.

danieljbruce committed:
Fix a test by not requiring the done callback

A test is timing out because we are waiting for done to be called. This fix does not require done to be called.

danieljbruce committed:
Merge branch 'main' into or-filters

danieljbruce committed:
Merge branch 'main' of https://github.com/googleapis/nodejs-datastore into or-filters

# Conflicts:
#	src/query.ts
#	test/query.ts

danieljbruce committed:
Revert "Fix a test by not requiring the done callback"

This reverts commit 1159b374f483d107736a765b5b4f1ade80116ce3.

danieljbruce committed:
Revert "Basic unit tests for setAncestor"

This reverts commit 86841d6e0a3d26e2df647fa4a42ed849b61b38a6.

danieljbruce committed:
Revert "Add setAncestor"

This reverts commit e84582faa6041429a4e81d47bd21dbc338a639e7.

danieljbruce committed:
Separate filters and new filters internally

This commit is done to avoid a breaking typescript change which could have the potential to affect some users who read `filters` on a query as its type had been changed to be more flexible, but is now back to what it was.

danieljbruce committed:
Move AND/OR into their own separate function

AND and OR should not be static functions of the filter class because then the user has to type Filter.AND instead of AND for example.

danieljbruce committed:
Eliminate unused imports

Artifacts of having imports laying around and moving functionality between files

danieljbruce commented on src/filter.ts:
```diff
@@ -0,0 +1,159 @@
+ * @see {@link https://cloud.google.com/datastore/docs/concepts/queries#filters| Filters Reference}
+ *
+ */
+export abstract class Filter {
```
Maybe this should be an interface.

danieljbruce committed:
Revert "Add a test for new warning"

This reverts commit bc15f325d87b4ffe46ae8727293c5cefa2300501.

danieljbruce committed:
Revert "Now removes deprecation warning properly"

This reverts commit db02a5089242b25b4bebae4304a4778f0bf6d17f.

danieljbruce commented on system-test/datastore.ts:
```diff
@@ -811,6 +812,83 @@ describe('Datastore', () => {
+          'Character',
+          'Eddard',
+        ]);
+        it('should run a query using hasAncestor first', async () => {
```
typo first/last should be switched around

danieljbruce replied:
Solved

Mariatta commented on system-test/datastore.ts:
```diff
@@ -811,6 +812,83 @@ describe('Datastore', () => {
+            ])
+          );
+        const [entities] = await datastore.runQuery(q);
+        assert.strictEqual(entities!.length, 6);
```
Is it possible to also assert that the resulting entities indeed have the values `family = Stark` and `apprearances >= 20`?

danieljbruce replied:
I looked at the data and there is actually one entity that has family `['Stark', 'Tully']`. I believe this is intended behaviour for the filter because this is also the result when we test a case where legacy `.filter` is used.

danieljbruce replied:
I just checked `should filter queries with defined indexes` on main. It seems this filter checks to see if `Stark` is present. I guess this matches what you said. We can check for its pressence.

danieljbruce replied:
Done in last commit

Mariatta commented on system-test/datastore.ts:
```diff
@@ -811,6 +812,83 @@ describe('Datastore', () => {
+            ])
+          );
+        const [entities] = await datastore.runQuery(q);
+        assert.strictEqual(entities!.length, 8);
```
Is it possible to assert that in resulting query, there are entities that have value `family = Stark` and not `appearances >=20`?

danieljbruce replied:
Right now I am seeing that all the entities which are returned have a family name of Stark.

danieljbruce replied:
Done in last commit

kolea2 commented on src/entity.ts:
```diff
@@ -1249,34 +1238,12 @@ export namespace entity {
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        let value: any = {};
-
-        if (filter.name === '__key__') {
```
why are we removing this `__key__` logic?

danieljbruce replied:
This OR query implementation created duplicate code so this code has not been removed, it has only been refactored into the `PropertyFilter` object.

Take a look at https://github.com/googleapis/nodejs-datastore/pull/1061/files#diff-87404c91a22dbabcc334f9a5288341281cb2b6625533e1207efcac16dc1e2ffdR77.

kolea2 commented on src/entity.ts:
```diff
@@ -1249,34 +1238,12 @@ export namespace entity {
+        filter => new PropertyFilter(filter.name, filter.op, filter.val)
+      );
+      const newFilters = query.newFilters;
+      queryProto.filter = AND(newFilters.concat(filters)).toProto();
```
a little confused by this, what is this logic doing? Can you add some comments?

danieljbruce replied:
Added comments to the code here.

This is mainly just a refactor. Old code is being moved into the `Filter` classes, but this new code does the same thing as the removed code. The functionality had to be pulled into the `Filter` classes to avoid creating duplicate code.

kolea2 commented on src/query.ts:
```diff
@@ -201,24 +207,29 @@ class Query {
    */
-  filter(property: string, value: {} | null): Query;
-  filter(property: string, operator: Operator, value: {} | null): Query;
+  filter(propertyOrFilter: string | NewFilter, value?: {} | null): Query;
```
why can't this just be `string | Filter`?

danieljbruce replied:
In this module, `NewFilter` is just what we call the imported `Filter` class. We cannot call it `Filter` because the name `Filter` is already taken in this module.

https://github.com/googleapis/nodejs-datastore/pull/1061/files#diff-ad0978c0145d3620d2a97bdff4307f9496529c0e3011a845e2d5bf1a169d2538R21

danieljbruce replied:
Filter has been renamed `entityFilter` so as not to cause confusion.

danieljbruce committed:
Modify test cases to capture nuances in data

We add additional asserts to the data in order to capture the nuances of the composite operator. For example, for the OR test we make sure the filter doesn’t always require both conditions to be true.

danieljbruce commented on src/query.ts:
```diff
@@ -76,6 +76,7 @@ class Query {
   namespace?: string | null;
   kinds: string[];
   filters: Filter[];
+  newFilters: NewFilter[];
```
At first I put all the filters into the `filters` object whether they were the old type of filter or the new filter class. This caused a compiler error however caught by https://github.com/googleapis/nodejs-datastore/pull/1067/files#diff-a9fb5a76e455ff1f3d63199cae37a32e289867ec02b1ec205b79468d074a211eR881.

Currently, users may be popping off the `filters` variable and expecting that variable to have the `val` property, but it won't if we allow `filter` to store the new `Filter` class.

danieljbruce committed:
Added comments to code that was refactored

Code for building the `filter` property of the query proto was pulled into the `Filter` object. Comments indicate how legacy functionality was maintained and which lines of code perform which task.

danieljbruce committed:
rename NewFilter to entity filter

rename new filter to entity filter to eliminate the need for an internal rename that causes confusion

danieljbruce committed:
Switch around last and first

These test cases have mistakes in their names. We should change them to reflect the position of `hasAncestor`.

kolea2 commented on src/query.ts:
```diff
@@ -170,7 +176,7 @@ class Query {
    * @see {@link https://cloud.google.com/datastore/docs/concepts/queries#datastore-property-filter-nodejs| Datastore Filters}
    *
-   * @param {string} property The field name.
+   * @param {string | NewFilter} propertyOrFilter The field name.
```
shouldn't this be EntityFilter now? (side note - how is this compiling?)

danieljbruce replied:
You are right. This needs to change. This is compiling because this is actually just a documentation comment.

danieljbruce replied:
Done

danieljbruce committed:
Change the comment to reflect the new name

The name of the base class is now EntityFilter. Adjust this comment so that it correctly matches the parameter type.

bcoe commented on src/entity.ts:
```diff
@@ -21,6 +21,7 @@ import {PathType} from '.';
 import {protobuf as Protobuf} from 'google-gax';
 import * as path from 'path';
 import {google} from '../protos/protos';
+import {AND, PropertyFilter} from './filter';
```
nit, upper case is usually reserved for variable constants `const FOO = 9;  BAR = 99`, to make it clear that this is an exported method, I might go with:

`and`/`or`,  or `applyAnd`, `applyOr`.

danieljbruce replied:
Choosing the former

bcoe commented on src/filter.ts:
```diff
@@ -0,0 +1,159 @@
+   */
+  // eslint-disable-next-line
+  toProto(): any {
+    const OP_TO_OPERATOR = new Map([
```
nit: I would probably hoist this to the top of the file, rather than declaring in `toProto`, at scale this will speed things up since you're allocating less `Maps`.

bcoe commented on src/query.ts:
```diff
@@ -201,24 +207,29 @@ class Query {
-      value = operatorOrValue as {};
-      operator = '=';
-    }
+    if (isFilter(propertyOrFilter)) {
```
This is good use of TypeScript's `is` functionality 👍

danieljbruce replied:
Thanks

bcoe reviewed (approved):
I left a few minor suggestions, this is looking really good IMO:

- A comprehensive suite of tests.
- Code reads well.

bcoe commented:
@danieljbruce the tile of this PR seems a bit off, I believe you're actually introducing the ability to filter using a class rather than just a string?

Perhaps a better title:

```
feat: introduce EntityFilter class with support for and/or filters
```

Or something like this.

danieljbruce committed:
rename and move constant map up

rename composite filter functions so that they don’t look like constants and move a map up so that it doesn’t have to be initialized every time.

kolea2 commented on src/entity.ts:
```diff
@@ -1249,34 +1238,20 @@ export namespace entity {
+      const filters = query.filters.map(
+        filter => new PropertyFilter(filter.name, filter.op, filter.val)
+      );
+      const newFilters = query.entityFilters;
```
nit - rename variable

danieljbruce replied:
This was done

danieljbruce committed:
Rename newFilters to entityFilters

danieljbruce committed:
Merge branch 'main' into or-filters
